src/Shared/ServerInfrastructure/HttpCharacters.cs

```diff
@@ -1,224 +1,91 @@
+
+    private static IndexOfAnyValues<char> CreateAllowedFieldChars()
     {
         // field-value https://tools.ietf.org/html/rfc7230#section-3.2
-        var fieldValue = new bool[_tableSize];
 
-        fieldValue[0x9] = true; // HTAB
+        Span<char> tmp = stackalloc char[128];
+        tmp[0] = (char)0x9;     // HTAB
+        var count = 1;
 
-        for (var c = 0x20; c <= 0x7e; c++) // VCHAR and SP
+        for (var c = 0x20; c <= 0x7E; ++c)  // VCHAR and SP
         {
-            fieldValue[c] = true;
+            tmp[count++] = (char)c;
         }
-        return fieldValue;
-    }
-
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public static bool ContainsInvalidAuthorityChar(Span<byte> s)
-    {
-        var authority = _authority;
 
-        for (var i = 0; i < s.Length; i++)
-        {
-            var c = s[i];
-            if (c >= (uint)authority.Length || !authority[c])
-            {
-                return true;
-            }
-        }
-
-        return false;
+        Debug.Assert(count <= tmp.Length);
+        return IndexOfAnyValues.Create(tmp[..count]);
     }
 
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public static int IndexOfInvalidHostChar(string s)
-    {
-        var host = _host;
-
-        for (var i = 0; i < s.Length; i++)
-        {
-            var c = s[i];
-            if (c >= (uint)host.Length || !host[c])
-            {
-                return i;
-            }
-        }
+    public static bool ContainsInvalidAuthorityChar(ReadOnlySpan<byte> span) => span.IndexOfAnyExcept(_allowedAuthorityBytes) >= 0;
 
-        return -1;
-    }
+    public static int IndexOfInvalidHostChar(ReadOnlySpan<char> span) => span.IndexOfAnyExcept(_allowedHostChars);
 
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public static int IndexOfInvalidTokenChar(string s)
-    {
-        var token = _token;
+    public static int IndexOfInvalidTokenChar(ReadOnlySpan<char> span) => span.IndexOfAnyExcept(_allowedTokenChars);
 
-        for (var i = 0; i < s.Length; i++)
-        {
-            var c = s[i];
-            if (c >= (uint)token.Length || !token[c])
-            {
-                return i;
-            }
-        }
+    public static int IndexOfInvalidTokenChar(ReadOnlySpan<byte> span) => span.IndexOfAnyExcept(_allowedTokenBytes);
 
-        return -1;
-    }
+    // Follows field-value rules in https://tools.ietf.org/html/rfc7230#section-3.2
+    // Disallows characters > 0x7E.
+    public static int IndexOfInvalidFieldValueChar(ReadOnlySpan<char> span) => span.IndexOfAnyExcept(_allowedFieldChars);
 
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public static int IndexOfInvalidTokenChar(ReadOnlySpan<byte> span)
+    // Follows field-value rules for chars <= 0x7F. Allows extended characters > 0x7F.
+    public static int IndexOfInvalidFieldValueCharExtended(ReadOnlySpan<char> span)
```

Comment:
You can use an inverted condition here.
Make a new
```c#
private static readonly IndexOfAnyValues<char> s_invalidFieldValueChars;
```
with this as the values (1-32 (except HTAB) + 0x7F)
```c#
"\u0000\u0001\u0002\u0003\u0004\u0005\u0006\u0007\u0008\u000A\u000B\u000C\u000D\u000E\u000F\u0010" +
"\u0011\u0012\u0013\u0014\u0015\u0016\u0017\u0018\u0019\u001A\u001B\u001C\u001D\u001E\u001F\u007F"
```

And have this be just
```c#
public static int IndexOfInvalidFieldValueCharExtended(ReadOnlySpan<char> span) => span.IndexOfAny(s_invalidFieldValueChars);
```

Reply:
🎉 thanks -- needing to 🐬 around in the brain, then it gets even more beautiful.